aeternity/oracle.py

```diff
@@ -0,0 +1,348 @@
+import json
+from collections import defaultdict, namedtuple
+
+import logging
+from json import JSONDecodeError
+
+import requests
+import time
+import websocket
+
+from aeternity.config import Config
+
+logger = logging.getLogger(__name__)
+
+
+class Connection:
+    def __init__(self, config):
+        self.config = config
+        self.websocket = None
+
+    def close(self):
+        self.websocket.close()
+        self.websocket = None
+
+    def assure_connected(self):
+        if self.websocket is None:
+            self.websocket = websocket.create_connection(self.config.websocker_url)
+
+    def receive(self):
+        self.assure_connected()
+        message = json.loads(self.websocket.recv())
+        logger.debug('received: %s', message)
+        return message
+
+    def send(self, message):
+        self.assure_connected()
+        logger.debug('sending: %s', message)
+        self.websocket.send(json.dumps(message))
+
+
+class EpochRequestError(Exception):
+    pass
+
+
+class EpochClient:
+    next_block_poll_interval_sec = 10
+
+    def __init__(self, config=None):
+        if config is None:
+            config = Config.get_default()
+        self._config = config
+        self._listeners = defaultdict(list)
+        self._connection = Connection(config=config)
+        self._top_block = None
+        # self.send_and_receive(
+        #     {"target":"chain", "action":"subscribe", "payload":{"type":"mined_block"}},
+        #     print
+        # )
+
+    def http_call(self, method, base_url, endpoint, **kwargs):
+        url = base_url + '/' + endpoint
+        response = requests.request(method, url, **kwargs)
+        if response.status_code >= 500:
+            raise EpochRequestError(response)
+        try:
+            return response.json()
+        except JSONDecodeError:
+            raise EpochRequestError(response.text)
+
+    def internal_http_get(self, endpoint, **kwargs):
+        return self.http_call(
+            'get', self._config.internal_api_url, endpoint, **kwargs
+        )
+
+    def internal_http_post(self, endpoint, **kwargs):
+        return self.http_call(
+            'post', self._config.internal_api_url, endpoint, **kwargs
+        )
+
+    def local_http_get(self, endpoint, **kwargs):
+        return self.http_call(
+            'get', self._config.http_api_url, endpoint, **kwargs
+        )
+
+    def local_http_post(self, endpoint, **kwargs):
+        return self.http_call(
+            'post', self._config.http_api_url, endpoint, **kwargs
+        )
+
+    def get_pubkey(self):
+        return self._config.get_pubkey()
+
+    def get_top_block(self):
+        data = requests.get(self._config.top_block_url).json()
+        return int(data['height'])
+
+    def update_top_block(self):
+        self._top_block = self.get_top_block()
+
+    def wait_for_next_block(self):
+        if self._top_block == None:
```

Comment:
I think a more generic method "wait for N blocks" with default set to 1 would be nicer.

Reply:
👍